clientv3/watch.go

```diff
@@ -251,13 +254,20 @@ func (w *watcher) run() {
 		// New events from the watch client
 		case pbresp := <-w.respc:
 			switch {
-			case pbresp.Canceled:
+			case !pbresp.Created && pbresp.Canceled:
 				delete(cancelSet, pbresp.WatchId)
-			case pbresp.Created:
+			case pbresp.Created && !pbresp.Canceled:
 				// response to pending req, try to add
 				w.addStream(pbresp, pendingReq)
 				pendingReq = nil
 				curReqC = w.reqc
+			case pbresp.Created && pbresp.Canceled:
```

Comment:
don't special case this; let `addStream` check for `Canceled` in the response message and then close the stream like it does for `Compacted` messages.